### PR TITLE
Optimize single-argument methods

### PR DESCRIPTION
**Pull request by JacobEvelyn**
This commit adds a performance optimization to one-argument
methods; these methods now have their memoization hashes
stored in an array rather than in an outer hash. This offers
a slight speedup because array accesses are faster than
hash lookups.

**Before merging:**

- [x] Copy the table printed at the end of the latest benchmark results into the `README.md` and update this PR
- [x] If this change merits an update to `CHANGELOG.md`, add an entry following Keep a Changelog [guidelines](https://keepachangelog.com/en/1.0.0/) with [semantic versioning](https://semver.org/)
- [x] Manually re-confirm that this code is consistently faster than the previous no-array implementation

### CODE REVIEW

File: spec/reset_memo_wise_spec.rb [hunk 1]
```diff
@@ -388,6 +388,48 @@
           expect(instance2.no_args_counter).to eq(1)
         end
       end
+
+      context "when method name is the same as a memoized class method" do
```
[the remark this answers is not in the document]
Comment (jemmaissroff) [reply spec/reset_memo_wise_spec.rb:392]: I think it makes sense to either leave a comment explaining this decision or to add similar tests to `preset_memo_wise_spec`, otherwise they're asymmetrical without a documented explanation for why.

File: spec/reset_memo_wise_spec.rb [hunk 2]
```diff
@@ -407,6 +491,52 @@
...
+                def one_arg(a) # rubocop:disable Naming/MethodParameterName
+                  @class_one_arg_counter = class_one_arg_counter + 1
+                  "class_one_arg: a=#{a}"
+                end
+                memo_wise :one_arg
+              end
+            end
+          end
+
+          it "resets memoization independently" do
```
Comment (jemmaissroff) [inline spec/reset_memo_wise_spec.rb:526]: Do you think it's overkill to also test this for the flip case - when memoization is reset on the instance method and not the class method?

**Review by jemmaissroff — CHANGES_REQUESTED** [review]

File: lib/memo_wise.rb [hunk 3]
```diff
@@ -179,73 +180,68 @@ def #{method_name}
...
+                output
+              else
+                hashes = (@_memo_wise_hashes[:#{method_name}] ||= Set.new)
+                hashes << key
+                @_memo_wise[key] = #{original_memo_wised_name}(#{MemoWise::InternalAPI.call_str(method)})
+              end
             end
-          end
+          END_OF_METHOD
+        else # :splat, :double_splat
```
[the remark this answers is not in the document]
Comment (JacobEvelyn) [reply lib/memo_wise.rb:230]: Because then we need to add a test that exercises the `else` case, which is impossible/very difficult because our cases are exhaustive. Does that make sense?

File: lib/memo_wise/internal_api.rb [hunk 4]
```diff
@@ -10,15 +10,32 @@ class InternalAPI
...
+      #   ]
+      # This is essentially a faster alternative to:
+      #   {
+      #     single_arg_method_name: { arg1 => :memoized_result, ... }
+      #   }
+      # because we can give each single-argument method its own array index at
+      # load time and perform that array lookup more quickly than a hash lookup
+      # by method name.
       obj.instance_variable_set(:@_memo_wise, {}) unless obj.instance_variables.include?(:@_memo_wise)
+      unless obj.instance_variables.include?(:@_memo_wise_single_argument)
```
[the remark this answers is not in the document]
Comment (JacobEvelyn) [reply lib/memo_wise/internal_api.rb:36]: I'd rather not do that now since at this point that hash also contains 0-arity methods, and I think it's possible that this optimization won't end up being faster for 0-arity methods because of the sentinel issue. Thoughts?

File: spec/support/shared_context_for_instance_methods.rb [hunk 5]
```diff
@@ -27,7 +27,36 @@ def protected_memowise_method
       protected :protected_memowise_method
       memo_wise :protected_memowise_method
 
-      # Counter for calls to class method '.positional_args', see below.
+      # Counter for calls to class method '.no_args', see below.
+      def self.class_no_args_counter
```
[the remark this answers is not in the document]
Comment (JacobEvelyn) [reply spec/support/shared_context_for_instance_methods.rb:31]: Done.

File: spec/reset_memo_wise_spec.rb (hunk 2 above)
Comment (JacobEvelyn) [reply spec/reset_memo_wise_spec.rb:526]: 👍 done

File: lib/memo_wise/internal_api.rb [hunk 6]
```diff
@@ -37,107 +54,81 @@ def self.create_memo_wise_state!(obj)
...
+    #   arguments it has
+    # @return [Symbol] one of:
+    #   - :none (example: `def foo`)
+    #   - :one_required_positional (example: `def foo(a)`)
+    #   - :one_required_keyword (example: `def foo(a:)`)
+    #   - :multiple_required (examples: `def foo(a, b)`, `def foo(a:, b:)`, `def foo(a, b:)`)
+    #   - :splat (examples: `def foo(a=1)`, `def foo(a, *b)`)
+    #   - :double_splat (examples: `def foo(a: 1)`, `def foo(a:, **b)`)
+    #   - :splat_and_double_splat (examples: `def foo(a=1, b: 2)`, `def foo(a=1, **b)`, `def foo(*a, **b)`)
+    def self.method_arguments(method)
```
[the remark this answers is not in the document]
Comment (JacobEvelyn) [reply lib/memo_wise/internal_api.rb:67]: 👍 done

File: lib/memo_wise/internal_api.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (jemmaissroff) [reply lib/memo_wise/internal_api.rb:36]: That makes sense to me

File: lib/memo_wise/internal_api.rb [hunk 7]
```diff
@@ -189,63 +180,13 @@ def initialize(target)
...
-        return true
-      end
-
-      MemoWise::InternalAPI.has_only_required_args?(method) &&
-        method.parameters.size > 1
+    # @param method_name [Symbol] the name of the memoized method
+    # @return [Integer] the array index in `@_memo_wise_single_argument` to use
+    #   to find the memoization data for the given method
+    def index(method_name)
+      indices = target_class.instance_variable_get(:@_memo_wise_indices) ||
```
[the remark this answers is not in the document]
Comment (jemmaissroff) [reply lib/memo_wise/internal_api.rb:195]: I don't know another way to do it either

**Review by jemmaissroff — APPROVED** [review]

File: lib/memo_wise.rb [hunk 8]
```diff
@@ -179,73 +180,68 @@ def #{method_name}
...
+          #   {
+          #     ClassBeingMemoized => {
+          #       single_arg_method_name: 0,
+          #       other_single_arg_method_name: 1
+          #     },
+          #     ClassBeingMemoizedSingletonClass => {
+          #       another_single_arg_method_name: 0
+          #     }
+          #   }
+          @_memo_wise_indices ||= Hash.new { |h, k| h[k] = {} }
```
[the remark this answers is not in the document]
Comment (JacobEvelyn) [reply lib/memo_wise.rb:196]: @ms-ati I'd love your thoughts on this as well.

File: lib/memo_wise/internal_api.rb (hunk 7 above)
[the remark this answers is not in the document]
Comment (JacobEvelyn) [reply lib/memo_wise/internal_api.rb:195]: @ms-ati I'd love your thoughts on this as well. I'm still worried about edge cases in which multiple `@_memo_wise_indices` instance vars exist on different objects and cause incorrect behavior when multiple memoized methods share the same name, but I'm not sure what additional tests I can add that I haven't already added.

File: spec/memo_wise_spec.rb [hunk 9]
```diff
@@ -195,11 +195,29 @@
       end
 
       context "with class method with same name as memoized instance method" do
-        it "does not memoize the class methods" do
-          expect(Array.new(4) { class_with_memo.with_positional_args(1, 2) }).
-            to all eq("class_with_positional_args: a=1, b=2")
+        context "for methods with no arguments" do
```
[the remark this answers is not in the document]
Comment (JacobEvelyn) [reply spec/memo_wise_spec.rb]: @ms-ati I'd love your thoughts on DRYing up these tests.

[the remark this answers is not in the document]
Comment (JacobEvelyn) [reply spec/memo_wise_spec.rb]: Never mind, I think I came up with a DRY approach in the `Round 2 changes` commit.

**Review by jemmaissroff — APPROVED** [review]

File: lib/memo_wise.rb (hunk 8 above)
[the remark this answers is not in the document]
Comment (ms-ati) [reply lib/memo_wise.rb:196]: @JacobEvelyn Is your reasoning that this is safe for Marshal, because the Hash with default proc lives in a Class vs in an instance? Do we have tests of the Marshal case that verify this?

File: lib/memo_wise/internal_api.rb (hunk 7 above)
[the remark this answers is not in the document]
Comment (ms-ati) [reply lib/memo_wise/internal_api.rb:195]: I agree that this doesn't seem like the correct pattern @JacobEvelyn. Should we schedule a pairing session to review the code, and look for ways to have a single code path that knows "where" to look?

I may be misunderstanding, but it seems like we should know whether we are in a "def",  "def self.", or "class << self" code path, shouldn't we?

File: lib/memo_wise.rb (hunk 8 above)
[the remark this answers is not in the document]
Comment (JacobEvelyn) [reply lib/memo_wise.rb]: My reasoning is more one of usage—I find it unlikely that someone would start loading a class (e.g. `memo_wise` one method), then Marshal it, then load it, and then `memo_wise` another method. But maybe I should just change this so we don't worry about it? Adding a test for it feels like it may be overkill though.

[the remark this answers is not in the document]
Comment (JacobEvelyn) [reply lib/memo_wise.rb]: nvm, I realized this was no longer needed and have removed it.

**Review by ms-ati — APPROVED** [review]
LGTM
